File: lib/services/mood_analyzer.dart
import 'dart:math';

/// App-side mood calculation from raw breath metrics.
/// Computes stress, focus, and meditation scores with configurable parameters.
class MoodAnalyzer {
  // Rolling window of recent breaths for calculations
  final List<BreathMetrics> _recentBreaths = [];
  static const int _maxRecent = 10;  // Increased for better smoothing
  
  // Calibration baseline (locked after calibration)
  double? _baselineCycle;  // Average cycle during calibration
  double? _baselinePeakFlow;  // Average peak flow during calibration
  bool _baselineLocked = false;
  
  // Successive differences for RMSSD calculation
  final List<double> _successiveDiffs = [];
  static const int _maxDiffs = 4;
  double? _lastCycleDuration;
  
  // Breath count for calibration
  int _breathCount = 0;
  
  // Current computed scores
  int _stressScore = 0;
  int _focusScore = 5;
  int _meditationScore = 0;
  
  // Current metrics
  double _currentRatio = 1.0;
  double _currentRmssd = 1.0;
  double _currentCycle = 4.0;
  double _currentConsistency = 0.5;
  
  // === CONFIGURABLE THRESHOLDS ===
  
  // Stress calculation
  double calmRatio = 1.5;           // E/I ratio at which stress is lowest
  double anxiousRatio = 0.8;        // E/I ratio at which stress is highest
  double calmVariabilityCv = 0.10;  // CV below this = calm
  double anxiousVariabilityCv = 0.40; // CV above this = stressed
  
  // Stress component weights (must sum to 1.0)
  double stressRatioWeight = 0.40;      // E/I ratio is primary indicator
  double stressDurationWeight = 0.25;    // Duration vs baseline
  double stressSmoothnessWeight = 0.05;  // Reduced - less reliable
  double stressPeakFlowWeight = 0.15;    // Shallow breaths = stressed
  double stressRmssdWeight = 0.15;
  
  // Focus calculation
  double focusConsistencyCv = 0.15; // CV threshold for focus
  
  // Meditation calculation
  double meditationMinCycle = 6.0;  // Below this = too fast
  double meditationOptimalCycle = 10.0; // Optimal cycle for meditation
  double meditationStabilityThreshold = 0.15;
  
  int calibrationBreaths = 6;
  
  bool get isCalibrating => _breathCount < calibrationBreaths;
  int get breathCount => _breathCount;
  
  // === SCORE GETTERS ===
  int get stressScore => _stressScore;
  int get focusScore => _focusScore;
  int get meditationScore => _meditationScore;
  
  // === RAW METRIC GETTERS (for expandable UI) ===
  double get currentRatio => _currentRatio;
  double get currentCycle => _currentCycle;
  double get currentConsistency => _currentConsistency;
  double get currentRmssd => _currentRmssd;
  double? get baselineCycle => _baselineCycle;
  double? get baselinePeakFlow => _baselinePeakFlow;
  bool get isBaselineLocked => _baselineLocked;
  
  // Latest smoothness and peak flow (from most recent breath)
  int get currentSmoothness => _recentBreaths.isEmpty ? 50 : _recentBreaths.last.smoothness;
  double get currentPeakFlow => _recentBreaths.isEmpty ? 0.0 : _recentBreaths.last.peakFlow;
  
  /// Process a new breath with raw metrics
  void recordBreath({
    required double exhaleDuration,
    required double inhaleDuration,
    required double cycleDuration,
    required int smoothness,
    double peakFlow = 0.0,
    int symmetry = 50,
  }) {
    // Validate breath (filter artifacts)
    if (exhaleDuration < 0.3 || inhaleDuration < 0.2) return;
    if (exhaleDuration > 15.0 || inhaleDuration > 10.0) return;
    
    _breathCount++;
    
    // Store recent breath
    _recentBreaths.add(BreathMetrics(
      exhaleDuration: exhaleDuration,
      inhaleDuration: inhaleDuration,
      cycleDuration: cycleDuration,
      smoothness: smoothness,
      peakFlow: peakFlow,
      symmetry: symmetry,
    ));
    if (_recentBreaths.length > _maxRecent) {
      _recentBreaths.removeAt(0);
    }
    
    // Track successive differences (RMSSD-like)
    if (_lastCycleDuration != null) {
      final diff = (cycleDuration - _lastCycleDuration!).abs();
      _successiveDiffs.add(diff);
      if (_successiveDiffs.length > _maxDiffs) {
        _successiveDiffs.removeAt(0);
      }
    }
    _lastCycleDuration = cycleDuration;
    
    // HYBRID ADAPTIVE BASELINE:
    // Initialize from calibration, then adapt SLOWLY (2% per breath) toward current patterns
    if (!_baselineLocked && _breathCount == calibrationBreaths && _recentBreaths.isNotEmpty) {
      // Initial baseline from calibration
      _baselineCycle = _recentBreaths.map((b) => b.cycleDuration).reduce((a, b) => a + b) / _recentBreaths.length;
      _baselinePeakFlow = _recentBreaths.map((b) => b.peakFlow).reduce((a, b) => a + b) / _recentBreaths.length;
      _baselineLocked = true;
    } else if (_baselineLocked) {
      // Slowly adapt baseline toward current values (2% adaptation rate - very slow)
      const adaptRate = 0.02;
      _baselineCycle = _baselineCycle! * (1 - adaptRate) + cycleDuration * adaptRate;
      if (peakFlow > 0 && _baselinePeakFlow != null) {
        _baselinePeakFlow = _baselinePeakFlow! * (1 - adaptRate) + peakFlow * adaptRate;
      }
    }
    
    // Compute all metrics and scores
    _updateMetrics();
    _computeStressScore(smoothness, peakFlow);
    _computeFocusScore();
    _computeMeditationScore();
  }
  
  void _updateMetrics() {
    if (_recentBreaths.isEmpty) return;
    
    // E/I Ratio (average over recent breaths)
    final ratios = _recentBreaths.map((b) => 
      b.inhaleDuration > 0 ? b.exhaleDuration / b.inhaleDuration : 1.0
    ).toList();
    _currentRatio = ratios.reduce((a, b) => a + b) / ratios.length;
    
    // RMSSD
    if (_successiveDiffs.length >= 2) {
      final squared = _successiveDiffs.map((d) => d * d).toList();
      _currentRmssd = sqrt(squared.reduce((a, b) => a + b) / squared.length);
    }
    
    // Consistency (std dev of breath durations)
    final durations = _recentBreaths.map((b) => b.cycleDuration).toList();
    if (durations.length >= 2) {
      final mean = durations.reduce((a, b) => a + b) / durations.length;
      final variance = durations.map((d) => pow(d - mean, 2)).reduce((a, b) => a + b) / durations.length;
      _currentConsistency = sqrt(variance);
    }
    
    // Average cycle duration
    _currentCycle = durations.reduce((a, b) => a + b) / durations.length;
  }
  
  void _computeStressScore(int smoothness, double peakFlow) {
    // 1. Ratio component: higher ratio = more calm = negative stress
    final ratioRange = calmRatio - anxiousRatio;
    final ratioNorm = (_currentRatio - anxiousRatio) / max(0.1, ratioRange);
    double ratioScore = 5 - (ratioNorm * 10);
    ratioScore = ratioScore.clamp(-5, 5);
    
    // 2. Duration component: compare to CALIBRATION baseline (not sliding window)
    double durationScore = 0;
    if (_baselineLocked && _baselineCycle != null) {
      final durationRatio = _currentCycle / max(0.1, _baselineCycle!);
      // <0.7 (faster than calibration) = stressed, >1.3 (slower) = calm
      final durationNorm = (durationRatio - 1.0) / 0.3;
      durationScore = (-durationNorm * 5).clamp(-5, 5);
    }
    
    // 3. Smoothness component: lower smoothness = more stress
    // smoothness 100 = score -5 (calm), smoothness 0 = score +5 (stressed)
    final smoothnessScore = (5 - (smoothness / 10)).clamp(-5.0, 5.0);
    
    // 4. Peak flow component: compare to CALIBRATION baseline
    double peakFlowScore = 0;
    if (_baselineLocked && _baselinePeakFlow != null && _baselinePeakFlow! > 0 && peakFlow > 0) {
      final peakRatio = peakFlow / _baselinePeakFlow!;
      // < 0.7 = shallow vs calibration = stressed, > 1.3 = deeper = calm
      peakFlowScore = ((peakRatio - 1.0) * 5).clamp(-5.0, 5.0);
    }
    
    // 5. RMSSD CV component
    final rmssdCv = _currentRmssd / max(0.1, _currentCycle);
    final rmssdRange = anxiousVariabilityCv - calmVariabilityCv;
    final rmssdNorm = (rmssdCv - calmVariabilityCv) / max(0.01, rmssdRange);
    double rmssdScore = -5 + (rmssdNorm * 10);
    rmssdScore = rmssdScore.clamp(-5, 5);
    
    // 6. ABSOLUTE SHORT BREATH PENALTY
    // Short breaths are stressful regardless of personal baseline
    double absoluteShortPenalty = 0;
    if (_currentCycle < 3.0) {
      absoluteShortPenalty = 3.0;  // Very short = add stress
    } else if (_currentCycle < 4.0) {
      absoluteShortPenalty = 2.0;  // Short = moderate stress
    } else if (_currentCycle < 5.0) {
      absoluteShortPenalty = 1.0;  // Slightly short = mild stress
    }
    
    // Combine with configurable weights
    final rawStress = stressRatioWeight * ratioScore +
                      stressDurationWeight * durationScore +
                      stressSmoothnessWeight * smoothnessScore +
                      stressPeakFlowWeight * peakFlowScore +
                      stressRmssdWeight * rmssdScore +
                      absoluteShortPenalty;
    
    _stressScore = rawStress.round().clamp(-5, 5);
  }
  
  void _computeFocusScore() {
    // Focus = inverse of consistency CV
    final consistencyCv = _currentConsistency / max(0.1, _currentCycle);
    
    // Low CV = high focus
    final cvRange = 0.30 - focusConsistencyCv;
    final cvNorm = (consistencyCv - focusConsistencyCv) / max(0.01, cvRange);
    final rawFocus = 10 - (cvNorm * 10);
    
    _focusScore = rawFocus.round().clamp(0, 10);
  }
  
  void _computeMeditationScore() {
    // Meditation based on cycle duration and stability
    
    // Duration component: 6-10s cycles are meditative
    double durationScore;
    if (_currentCycle < meditationMinCycle) {
      durationScore = 0; // Too fast
    } else if (_currentCycle >= meditationOptimalCycle) {
      durationScore = 10; // Optimal or slower
    } else {
      // Linear interpolation
      durationScore = 10 * (_currentCycle - meditationMinCycle) / (meditationOptimalCycle - meditationMinCycle);
    }
    
    // Stability bonus: consistent breathing deepens meditation
    final consistencyCv = _currentConsistency / max(0.1, _currentCycle);
    double stabilityMultiplier;
    if (consistencyCv <= meditationStabilityThreshold) {
      stabilityMultiplier = 1.0; // Very stable
    } else if (consistencyCv >= 0.4) {
      stabilityMultiplier = 0.5; // Unstable
    } else {
      stabilityMultiplier = 1.0 - ((consistencyCv - meditationStabilityThreshold) / 0.25) * 0.5;
    }
    
    final rawMeditation = durationScore * stabilityMultiplier;
    _meditationScore = rawMeditation.round().clamp(0, 10);
  }
  
  /// Reset analyzer state (for new session)
  void reset() {
    _recentBreaths.clear();
    _successiveDiffs.clear();
    _lastCycleDuration = null;
    _breathCount = 0;
    _stressScore = 0;
    _focusScore = 5;
    _meditationScore = 0;
    // Clear calibration baselines for next session
    _baselineCycle = null;
    _baselinePeakFlow = null;
    _baselineLocked = false;
  }
  
  /// Get stress level label for display
  String get stressLabel {
    if (isCalibrating) return 'Calibrating...';
    if (_stressScore <= -4) return 'SERENE';
    if (_stressScore <= -2) return 'CALM';
    if (_stressScore <= 1) return 'BALANCED';
    if (_stressScore <= 3) return 'TENSE';
    return 'ANXIOUS';
  }

  /// Get focus level label for display
  String get focusLabel {
    if (isCalibrating) return 'Calibrating...';
    if (_focusScore >= 8) return 'DEEP FOCUS';
    if (_focusScore >= 6) return 'FOCUSED';
    if (_focusScore >= 4) return 'ATTENTIVE';
    if (_focusScore >= 2) return 'DISTRACTED';
    return 'SCATTERED';
  }

  /// Get meditation depth label for display
  String get meditationLabel {
    if (isCalibrating) return 'Calibrating...';
    if (_meditationScore >= 8) return 'DEEP';
    if (_meditationScore >= 6) return 'MEDITATIVE';
    if (_meditationScore >= 4) return 'RELAXED';
    if (_meditationScore >= 2) return 'SETTLING';
    return 'ACTIVE';
  }
}

/// A single breath's raw metrics
class BreathMetrics {
  final double exhaleDuration;
  final double inhaleDuration;
  final double cycleDuration;
  final int smoothness;
  final double peakFlow;
  final int symmetry;
  
  BreathMetrics({
    required this.exhaleDuration,
    required this.inhaleDuration,
    required this.cycleDuration,
    required this.smoothness,
    this.peakFlow = 0.0,
    this.symmetry = 50,
  });
}

File: firmware/mood_analyzer.py
# mood_analyzer.py - Mood/State Detection from Breath Patterns
# Computes stress score (-5 serene to +5 anxious), focus level (0-10),
# and meditation depth (0-10) from thermistor breath data.
#
# SCIENTIFIC BASIS:
# 1. Respiratory Sinus Arrhythmia (RSA): Longer exhales activate parasympathetic
# 2. Heart Rate Variability proxied by breath variability: Lower HRV = more stress
# 3. Attentional control: Consistent timing = focused state
# 4. Mindfulness literature: Slow, rhythmic breathing (6 breaths/min) = meditative

import math

# === STRESS DETECTION ===
# Based on exhale/inhale ratio and breath-to-breath consistency
# Research: E/I ratio > 1.0 correlates with parasympathetic activation
STRESS_RATIO_ANXIOUS = 0.8    # Short exhale relative to inhale = stressed
STRESS_RATIO_CALM = 1.5       # Long exhale = calm (1.5x inhale length)

# Successive breath difference (RMSSD analog for breathing)
# Large jumps between consecutive breaths = stress response
STRESS_RMSSD_ANXIOUS = 2.0    # >2 second jumps between breaths = stressed  
STRESS_RMSSD_CALM = 0.5       # <0.5 second variation = very calm

# === FOCUS DETECTION ===
# Based on consistency of breath timing (like HRV for attention)
# Research: Focused attention produces regular, consistent breathing
FOCUS_CONSISTENCY_THRESHOLD = 0.5  # Std dev of last 5 breaths in seconds (was 0.3)
FOCUS_DRIFT_THRESHOLD = 1.0        # Max acceptable drift from target rhythm

# === MEDITATION DETECTION ===
# Based on breath cycle duration and stability
# Research: 4-6 breaths/min (10-15s cycles) = resonance breathing
MEDITATION_CYCLE_MIN = 6.0    # Below 6s = too fast for meditation
MEDITATION_CYCLE_OPTIMAL = 10.0  # 6 breaths/min = HRV resonance
MEDITATION_STABILITY_THRESHOLD = 0.15  # CV below this = stable rhythm

# Calibration
CALIBRATION_BREATHS = 6  # Reduced for faster feedback


class MoodAnalyzer:
    """Analyzes breath patterns using evidence-based metrics."""

    def __init__(self):
        self.breath_count = 0
        self.is_calibrating = True

        # Configurable thresholds (can be updated via set_thresholds)
        self.calm_ratio = STRESS_RATIO_CALM
        self.calm_variability = STRESS_RMSSD_CALM
        self.focus_consistency = FOCUS_CONSISTENCY_THRESHOLD
        self.calibration_breaths = CALIBRATION_BREATHS

        # Rolling window of recent breaths (short for responsiveness)
        self.recent_breaths = []  # List of (exhale_dur, inhale_dur, total_dur)
        self.max_recent = 5       # Only last 5 breaths for quick response
        
        # Successive difference tracking (RMSSD-like)
        self.last_breath_duration = None
        self.successive_diffs = []
        self.max_diffs = 4
        
        # Current computed metrics
        self.current_ratio = 1.0
        self.current_rmssd = 1.0  # Root mean square of successive differences
        self.current_consistency = 0.5  # Std dev of recent breath durations
        self.current_cycle = 4.0  # Average breath cycle duration

        # Scores
        self._stress_score = 0
        self._focus_score = 5
        self._meditation_score = 0

    def set_thresholds(self, calm_ratio, calm_variability, focus_consistency, calibration_breaths):
        """Update configurable thresholds from settings."""
        self.calm_ratio = calm_ratio
        self.calm_variability = calm_variability
        self.focus_consistency = focus_consistency
        self.calibration_breaths = int(calibration_breaths)
        print("[MOOD] Thresholds: ratio={} var={} focus={} cal={}".format(
            calm_ratio, calm_variability, focus_consistency, calibration_breaths
        ))

    def start_exhale(self):
        """Called when exhale begins."""
        pass  # Not needed with new algorithm

    def record_sample(self, flow_value):
        """Record exhale sample (limited to prevent memory issues)."""
        pass  # Not needed - using duration-based analysis

    def end_exhale(self, exhale_duration, inhale_duration, now):
        """Process completed breath cycle."""
        # Validate breath (filter artifacts)
        if exhale_duration < 0.3 or inhale_duration < 0.2:
            return  # Too short, likely artifact
        if exhale_duration > 15.0 or inhale_duration > 10.0:
            return  # Too long, likely pause/hold
            
        total_duration = exhale_duration + inhale_duration
        self.breath_count += 1
        
        # Store recent breath
        self.recent_breaths.append((exhale_duration, inhale_duration, total_duration))
        if len(self.recent_breaths) > self.max_recent:
            self.recent_breaths.pop(0)
        
        # Track successive differences (RMSSD-like)
        if self.last_breath_duration is not None:
            diff = abs(total_duration - self.last_breath_duration)
            self.successive_diffs.append(diff)
            if len(self.successive_diffs) > self.max_diffs:
                self.successive_diffs.pop(0)
        self.last_breath_duration = total_duration
        
        # Compute all metrics
        self._update_metrics()
        
        # Compute scores
        self._compute_stress_score()
        self._compute_focus_score()
        self._compute_meditation_score()
        
        # End calibration after enough breaths
        if self.breath_count >= self.calibration_breaths:
            self.is_calibrating = False

    def _update_metrics(self):
        """Update derived metrics from recent breaths."""
        if not self.recent_breaths:
            return
            
        # Exhale/Inhale ratio (averaged over recent breaths)
        ratios = [ex / max(0.1, inh) for ex, inh, _ in self.recent_breaths]
        self.current_ratio = sum(ratios) / len(ratios)
        
        # RMSSD (root mean square of successive differences)
        if len(self.successive_diffs) >= 2:
            squared = [d * d for d in self.successive_diffs]
            self.current_rmssd = math.sqrt(sum(squared) / len(squared))
        
        # Consistency (std dev of breath durations)
        durations = [total for _, _, total in self.recent_breaths]
        if len(durations) >= 2:
            mean_dur = sum(durations) / len(durations)
            variance = sum((d - mean_dur) ** 2 for d in durations) / len(durations)
            self.current_consistency = math.sqrt(variance)
        
        # Average cycle duration
        self.current_cycle = sum(durations) / len(durations)

    def _compute_stress_score(self):
        """
        Stress: -5 (serene) to +5 (anxious)
        
        Based on:
        1. E/I Ratio: Low ratio (short exhale) = stress (sympathetic)
        2. RMSSD: High variability between successive breaths = stress
        """
        # Ratio component: higher ratio = more calm = negative stress
        ratio_range = self.calm_ratio - STRESS_RATIO_ANXIOUS
        ratio_norm = (self.current_ratio - STRESS_RATIO_ANXIOUS) / max(0.1, ratio_range)
        ratio_score = 5 - (ratio_norm * 10)  # Invert: high ratio = low stress
        ratio_score = max(-5, min(5, ratio_score))
        
        # RMSSD component: high RMSSD = erratic = stressed
        rmssd_range = STRESS_RMSSD_ANXIOUS - self.calm_variability
        rmssd_norm = (self.current_rmssd - self.calm_variability) / max(0.1, rmssd_range)
        rmssd_score = -5 + (rmssd_norm * 10)  # High RMSSD = high stress
        rmssd_score = max(-5, min(5, rmssd_score))
        
        # Combine: 70% ratio (primary), 30% RMSSD (secondary)
        raw_stress = 0.7 * ratio_score + 0.3 * rmssd_score
        self._stress_score = int(round(max(-5, min(5, raw_stress))))
        
        # Debug: log first few scores
        if self.breath_count <= 10:
            print("[STRESS] ratio={:.2f} rmssd={:.2f} -> ratio_sc={:.1f} rmssd_sc={:.1f} -> {}".format(
                self.current_ratio, self.current_rmssd, ratio_score, rmssd_score, self._stress_score
            ))

    def _compute_focus_score(self):
        """
        Focus: 0 (distracted) to 10 (deeply focused)
        
        Based on breath consistency (std dev of timing).
        Focused attention = regular, consistent breathing rhythm.
        """
        # Low consistency (low std dev) = high focus
        thresh = self.focus_consistency
        if self.current_consistency < thresh * 0.5:
            # Very consistent = high focus
            focus = 9 + min(1, (thresh * 0.5 - self.current_consistency) * 5)
        elif self.current_consistency < thresh:
            # Moderately consistent
            norm = self.current_consistency / thresh
            focus = 6 + (1 - norm) * 3  # 6-9 range
        elif self.current_consistency < thresh * 2:
            # Some variation
            norm = (self.current_consistency - thresh) / thresh
            focus = 3 + (1 - norm) * 3  # 3-6 range
        else:
            # Highly variable = distracted
            focus = max(0, 3 - (self.current_consistency - thresh * 2))
        
        self._focus_score = int(round(max(0, min(10, focus))))
        
        # Debug: log first few
        if self.breath_count <= 10:
            print("[FOCUS] consistency={:.2f}s -> score={}".format(
                self.current_consistency, self._focus_score
            ))

    def _compute_meditation_score(self):
        """
        Meditation: 0 (active) to 10 (deep meditation)
        
        Based on:
        1. Breath cycle duration (longer = deeper)
        2. Rhythm stability (consistent = meditative)
        """
        if len(self.recent_breaths) < 3:
            self._meditation_score = 0
            return
        
        # Cycle duration component (50% weight)
        if self.current_cycle < MEDITATION_CYCLE_MIN:
            # Too fast
            cycle_score = (self.current_cycle / MEDITATION_CYCLE_MIN) * 4
        elif self.current_cycle < MEDITATION_CYCLE_OPTIMAL:
            # Good range, increasing
            progress = (self.current_cycle - MEDITATION_CYCLE_MIN) / (MEDITATION_CYCLE_OPTIMAL - MEDITATION_CYCLE_MIN)
            cycle_score = 4 + progress * 6
        else:
            # At or above optimal
            cycle_score = 10
        
        # Stability component (50% weight)
        cv = self.current_consistency / max(0.1, self.current_cycle)  # Coefficient of variation
        if cv < MEDITATION_STABILITY_THRESHOLD:
            stability_score = 10
        elif cv < MEDITATION_STABILITY_THRESHOLD * 2:
            norm = (cv - MEDITATION_STABILITY_THRESHOLD) / MEDITATION_STABILITY_THRESHOLD
            stability_score = 10 - norm * 5
        else:
            stability_score = max(0, 5 - (cv - MEDITATION_STABILITY_THRESHOLD * 2) * 10)
        
        # Combine
        raw_med = 0.5 * cycle_score + 0.5 * stability_score
        self._meditation_score = int(round(max(0, min(10, raw_med))))

    def get_stress_score(self):
        """Get stress score or None if calibrating."""
        return None if self.is_calibrating else self._stress_score

    def get_focus_score(self):
        """Get focus score or None if calibrating."""
        return None if self.is_calibrating else self._focus_score

    def get_meditation_score(self):
        """Get meditation score or None if calibrating.""" 
        return None if self.is_calibrating else self._meditation_score

    def get_calibrating(self):
        """Returns True if still calibrating."""
        return self.is_calibrating

    def get_calibration_progress(self):
        """Returns calibration progress 0.0 to 1.0."""
        return min(1.0, self.breath_count / CALIBRATION_BREATHS)
